README.md
## Introduction

[![Join the chat at https://gitter.im/oklai/koala](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/oklai/koala?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[http://koala-app.com](http://koala-app.com)

Koala is a GUI application for LESS, Sass and CoffeeScript compilation, to help web developers use these tools more efficiently for development.

[项目中文主页](http://koala-app.com/index-zh.html)

## Downloads 
v2.1.0 release [Changelog](https://github.com/oklai/koala/blob/master/Changelog.md)

[Download](http://koala-app.com) 


## Features

* **Multi-language Support:** support for LESS, Sass, Compass Framework, CoffeeScript and Dust.
* **Real-time Compilation:** listens to files, compiles automatically when file changes occur. Everything runs in the background without the need for user input.
* **Compile Options Support:** compilation options are avaliable per file.
* **Compression:** automatic compression of code after compilation has completed.
* **Error Notification:** if an error is encountered during compilation, Koala will display an error notification.
* **Cross-platform:** Koala runs perfectly on Windows, Linux and Mac.

## Screenshots

![linux](http://oklai.github.com/koala/img/screenshots/linux.png)

## Documents

`koala` is built with [node-webkit](https://github.com/rogerwang/node-webkit). node-webkit is an app runtime based on Chromium and node.js. You can write native apps in HTML and Javascript with node-webkit. 

### How to run Koala source code?
1. Clone Koala to the local;
2. For windows, **I DON'T KNOW YET**
For Mac OS X and Linux, run `./install.sh`
3. Download [node-webkit](https://github.com/rogerwang/node-webkit) prebuilt binaries for your system environment.
4. For windows, copy `nw.exe, nw.pak, icudt.dll` to `koala/src` directory, and install [Ruby](http://www.ruby-lang.org/) to `koala/src/ruby`;  
For Linux, copy `nw, nw.pak`, install Ruby: `sudo apt-get install ruby`;  
For Mac OS X, copy `node-webkit.app`, OS X already have Ruby installed.
5. Run the `nw` executable file.

## License

`koala`'s code uses the Apache license, Version 2.0, see our `LICENSE` file.


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/oklai/koala/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



Changelog.md
## 2.3.0
2017-11-19

* Enhance the windows installer (fixes #638)
* Update autoprefix to v7.1.6
* Update clean-css to v4.1.9
* Update coffee-script to v1.12.7
* Update postcss to v6.0.14
* Update uglify-js to v3.1.9
* Update uglify-es to v3.1.9
* Update sass to v3.5.2 (fixes #697)
* Replace setting GEM_HOME and GEM_PATH with Gem.paths. (fixes #679)
* Add support for distinguishing UglifyJS errors from bugs. (fixes #645)
* Add Polish language. (Thanks to @djmassive) (#693)
* Fix French translation. (Thanks to @NumEricR) (#485, #486)
* Add Persian language. (Thanks to @mw2017) (#494)
* Add Korean language. (Thanks to @k22pr) (#660)

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.3.0/KoalaSetup.exe)
* [macOS](https://github.com/oklai/koala/releases/download/v2.3.0/Koala.dmg)
* [Linux 32 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_i386.tar.gz)
* [Linux 64 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_x86_64.tar.gz)
* [Debian 32 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_i386.deb)
* [Debian 64 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_x86_64.deb)
* [RPM 32 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_i386.rpm)
* [RPM 64 bit](https://github.com/oklai/koala/releases/download/v2.3.0/koala_2.3.0_x86_64.rpm)

## 2.2.0
2017-05-14

* Update autoprefixer to v7.0.1 and clean-css to v4.1.2
* Update uglify-js to v3.0.4 and add an option to use ES6+. (fixes #622)
* Add option to disable auto-compiling files when project is added or reloaded. (fixes #367)
* Add support for resizing the main Koala window. (fixes #187)
* Force the Ruby string encoding to UTF-8 when executing sass and compass.
* Prevent the notification window from getting focused when shown. (fixes #621)

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.2.0/KoalaSetup.exe)
* [macOS](https://github.com/oklai/koala/releases/download/v2.2.0/Koala.dmg)
* [Linux 32 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_i386.tar.gz)
* [Linux 64 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_x86_64.tar.gz)
* [Debian 32 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_i386.deb)
* [Debian 64 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_x86_64.deb)
* [RPM 32 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_i386.rpm)
* [RPM 64 bit](https://github.com/oklai/koala/releases/download/v2.2.0/koala_2.2.0_x86_64.rpm)

## 2.1.4
2017-05-06

* Correctly prevent hiding the app icon in the dock on mac after a notification shown. (fixes #608)
* Fix CSS compiler. (fixes #611, fixes #615)

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.1.4/KoalaSetup.exe)
* [macOS](https://github.com/oklai/koala/releases/download/v2.1.4/Koala.dmg)
* [Linux 32 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_i386.tar.gz)
* [Linux 64 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_x86_64.tar.gz)
* [Debian 32 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_i386.deb)
* [Debian 64 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_x86_64.deb)
* [RPM 32 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_i386.rpm)
* [RPM 64 bit](https://github.com/oklai/koala/releases/download/v2.1.4/koala_2.1.4_x86_64.rpm)

## 2.1.3
2017-05-02

* Fix the CSS compiler. (fixes #607)
* Prevent hiding the app icon in the dock on mac after a notification shown. (fixes #608)
* Add support for quitting using cmd+q on mac. (fixes #571, fixes #348)
* Increase the timeout and the max buffer size for the sass and compass exec commands. (refs #466)
* Remove unneeded files from the app package.

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.1.3/KoalaSetup.exe)
* [macOS](https://github.com/oklai/koala/releases/download/v2.1.3/Koala.dmg)
* [Linux 32 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_i386.tar.gz)
* [Linux 64 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_x86_64.tar.gz)
* [Debian 32 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_i386.deb)
* [Debian 64 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_x86_64.deb)
* [RPM 32 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_i386.rpm)
* [RPM 64 bit](https://github.com/oklai/koala/releases/download/v2.1.3/koala_2.1.3_x86_64.rpm)

## 2.1.2
2017-05-01

* Add support for choosing install dir on windows. (refs #596)
* Add support for adding Koala to the “Programs and Features” in windows. (refs #605)
* Prevent zipping the app data on windows. (fixes #578, #596, #598, #605)
* Update the clean-css, coffee-script and uglify-js.

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.1.2/KoalaSetup.exe)
* [macOS](https://github.com/oklai/koala/releases/download/v2.1.2/Koala.dmg)
* [Linux 32 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_i386.tar.gz)
* [Linux 64 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_x86_64.tar.gz)
* [Debian 32 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_i386.deb)
* [Debian 64 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_x86_64.deb)
* [RPM 32 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_i386.rpm)
* [RPM 64 bit](https://github.com/oklai/koala/releases/download/v2.1.2/koala_2.1.2_x86_64.rpm)

## 2.1.1
2017-04-27

* Fix the notification window. (fixes #579)
* Fix autoprefix. (fixes #588, #584)
* Pin nwjs version to 0.21.6. (fixes #580)

Download:
* [Windows](https://github.com/oklai/koala/releases/download/v2.1.1/KoalaSetup.exe)

gulpfile.js
var path = require('path');
var gulp = require('gulp');
var shelljs = require('shelljs');
var mergeStream = require('merge-stream');
var runSequence = require('run-sequence');
var manifest = require('./src/package.json');
var NwBuilder = require('nw-builder');

var srcDir = path.resolve('./src');
var nwOptions = {
  version: 'latest',
  flavor: 'normal',
  appName: 'Koala',
  appVersion: manifest.version,
};

// Build for macOS
gulp.task(`build:osx64`, function(callback) {
  var nw = new NwBuilder(Object.assign(nwOptions, {
    files: [
      `${srcDir}/**`,
      `!${srcDir}/app/assets/{less,test}/**`,
      `!${srcDir}/app/templates/**`,
      `!${srcDir}/ruby/**`,
    ],
    platforms: ['osx64'],
    macIcns: path.resolve('./assets-osx/icon.icns'),
    macPlist: {
      NSHumanReadableCopyright: 'koala-app.com',
      CFBundleIdentifier: 'com.koala-app.koala',
    },
  }));

  nw.on('log', console.log.bind(console));
  nw.build(callback);
});

// Build for linux
[32, 64].forEach(function(arch) {
  gulp.task(`build:linux${arch}`, function(callback) {
    var nw = new NwBuilder(Object.assign(nwOptions, {
      files: [
        `${srcDir}/**`,
        `!${srcDir}/app/assets/{less,test}/**`,
        `!${srcDir}/app/templates/**`,
        `!${srcDir}/ruby/**`,
      ],
      platforms: [`linux${arch}`],
    }));

    nw.on('log', console.log.bind(console));
    nw.build(callback);
  });
});

// Build for win32
// On macOS/Linux, you need Wine installed to build win32 (or remove winIco below)
gulp.task(`build:win32`, function(callback) {
  var nw = new NwBuilder(Object.assign(nwOptions, {
    files: [
      `${srcDir}/**`,
      `!${srcDir}/app/assets/{less,test}/**`,
      `!${srcDir}/app/templates/**`,
      `!${srcDir}/ruby/{include,share}/**`,
      `!${srcDir}/ruby/lib/{pkgconfig,tcltk}/**`,
    ],
    platforms: ['win32'],
    winIco: process.argv.indexOf('--noicon') > 0 ? null : path.resolve('./assets-windows/icon.ico'),
    zip: false,
  }));

  nw.on('log', console.log.bind(console));
  nw.build(callback);
});

// Only runs on macOS (requires XCode properly configured)
gulp.task('sign:osx64', ['build:osx64'], function() {
  // shelljs.exec('codesign -v -f -s "Alexandru Rosianu Apps" ./build/Koala/osx64/Koala.app/Contents/Frameworks/*');
  // shelljs.exec('codesign -v -f -s "Alexandru Rosianu Apps" ./build/Koala/osx64/Koala.app');
  // shelljs.exec('codesign -v --display ./build/Koala/osx64/Koala.app');
  // shelljs.exec('codesign -v --verify ./build/Koala/osx64/Koala.app');
});

// Create a DMG for osx64
// Only works on macOS because of appdmg
gulp.task('pack:osx64', ['sign:osx64'], function() {
  shelljs.mkdir('-p', './dist');            // appdmg fails if ./dist doesn't exist
  shelljs.rm('-f', './dist/Koala.dmg');   // appdmg fails if the dmg already exists

  return gulp.src([])
    .pipe(require('gulp-appdmg')({
      source: './assets-osx/dmg.json',
      target: './dist/Koala.dmg',
    }));
});

// Create a NSIS installer for win32
// Must have `makensis` installed
gulp.task('pack:win32', ['build:win32'], function() {
  shelljs.exec('makensis ./assets-windows/installer.nsi');
});

// Create packages for linux
[32, 64].forEach(function(arch) {
  gulp.task(`prepare:linux${arch}`, [`build:linux${arch}`], function() {
    shelljs.rm('-rf', `./build/linux${arch}`);

    const move_data = gulp.src([
      './assets-linux/koala.desktop',
      `./build/Koala/linux${arch}/**`,
    ])
      .pipe(gulp.dest(`./build/linux${arch}/usr/share/koala`));

    const move_icons = gulp.src('./assets-linux/icons/**')
      .pipe(gulp.dest(`./build/linux${arch}/usr/share/icons`));

    return mergeStream(move_data, move_icons);
  });

  ['deb', 'rpm'].forEach(function(target) {
    gulp.task(`pack:linux${arch}:${target}`, [`prepare:linux${arch}`], function() {
      const port = arch == 32 ? 'i386' : 'x86_64';

      shelljs.mkdir('-p', './dist'); // it fails if the dir doesn't exist
      shelljs.exec("'" + [
        'fpm',
        '--output-type', target,
        '--input-type', 'dir',
        '--chdir', `./build/linux${arch}`,
        '--package', `./dist/koala_${manifest.version}_${port}.${target}`,
        '--force',
        '--name', 'koala',
        '--version', manifest.version,
        '--license', 'Apache',
        '--vendor', '',
        '--category', 'devel',
        '--depends', 'ruby >= 1.8.7',
        '--architecture', port,
        '--maintainer', 'Ethan Lai <lain.z.q@gmail.com>',
        '--description', 'A cool tool for web developers.\nKoala is a GUI application for Less, Sass, Compass and CoffeeScript compilation, to help web developers to use them more efficient.',
        '--url', 'http://koala-app.com',
        '--after-install', './assets-linux/after-install.sh',
        '--after-remove', './assets-linux/after-remove.sh',
        '--deb-priority', 'optional',
        '--rpm-os', 'linux',
        '.',
      ].join("' '") + "'");
    });
  });

  gulp.task(`pack:linux${arch}:targz`, [`prepare:linux${arch}`], function() {
    const port = arch == 32 ? 'i386' : 'x86_64';

    shelljs.mkdir('-p', './dist'); // it fails if the dir doesn't exist
    shelljs.exec(`tar -czf './dist/koala_${manifest.version}_${port}.tar.gz' -C './build/linux${arch}' .`);
  });
});


// Make packages for all platforms
gulp.task('pack:all', function(callback) {
  runSequence('pack:osx64', 'pack:win32', 'pack:linux', callback);
});

// Make packages for linux platforms
gulp.task('pack:linux', function(callback) {
  runSequence('pack:linux32:deb', 'pack:linux64:deb', 'pack:linux32:rpm', 'pack:linux64:rpm', 'pack:linux32:targz', 'pack:linux64:targz', callback);
});

// Build osx64 and run it
gulp.task('run:osx64', ['build:osx64'], function() {
  shelljs.exec('open ./build/Koala/osx64/Koala.app');
});

// Run osx64 without building
gulp.task('open:osx64', function() {
  shelljs.exec('open ./build/Koala/osx64/Koala.app');
});

// Make packages for all platforms by default
gulp.task('default', ['pack:all']);
